spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,343 @@
+
+[source,text]
+----
+$ pkl eval <module> --external-resource-reader ldap=pkl-ldap
+username = "john_appleseed"
+email = "appleseed@example.com"
+----
+
+In this example, the external reader may provide both `ldap:` and `ldaps:` schemes.
+External readers are registered by their scheme, so to support both schemes both would need to be specified on the command line:
+[source,text]
+----
+$ pkl eval <module> --external-resource-reader ldap=pkl-ldap --external-resource-reader ldaps=pkl-ldap
+----
+
+Registering an external reader for a scheme automatically adds that scheme to the default allowed modules/resources.
+As with Pkl's built-in module and resource schemes, setting explicit allowed module or resources overrides this behavior and appropriate patterns must be specified to allow use of external readers.
+
+== Detailed design
+
+To avoid terminology confusion with the existing language binding message passing model, the `pkl` process will remain known as the "server" and the reader process will be known as the "client", despite all requests in this proposal being initiated by the `pkl` process.
+
+=== Overall Flow
+
+1. User requests an evaluator (via Java API, Message Passing API, or CLI) with an external reader.
+2. The entrypoint (client Java code, `CliEvaluator`, or `Server`) instantiates an `ExternalProcess` based on the configured external readers.
+3. The entrypoint instantiates resource readers (`ResourceReaders.External`) and module key factories (`ModuleKeyFactories.External`) referencing the `ExternalProcess` and uses them to build an evaluator.
+`ResourceReaders.External` and `ModuleKeyFactories.External` function almost identically to the existing `ClientResourceReader` and `ClientModuleKeyFactory` and pkl-server will be updated to reuse these implementations where possible.
+4. Evaluation begins.
+5. A resource or module read is encountered.
+6. Upon first read only:
+1.. The `ExternalProcess` is started, which starts the configured child process.
+2.. The child process is sent a `Discover(Module|Resource)ReaderRequest` message.
+3.. A `DiscoverReadersResponse` message is received (within some timeout).
+7. The appropriate `Read*Request` message is sent to the child process, the corresponding `Read*Response` message is awaited.
+8. Evaluation continues.
+9. Evaluation completes.
+10. The `ExternalProcess` is closed, the child process is sent SIGTERM.
+11. If the child process has not terminated after some timeout (3 seconds?), it is forcefully stopped via SIGKILL.
+
+=== Java API
+
+New APIs:
+
+* `ResourceReaders.External` - a `ResourceReader` implementation identical to `ClientResourceReader`.
+* `ModuleKeyFactories.External` - a `ModuleKeyFactory` implementation similar to `ClientModuleKeyFactory`.
+* `ModuleKeys.External` - a `ModuleKey` implementation identical to `ClientModuleKey`.
```

Comment:
Updated to note exactly which parts of pol-server can be moved, ported, and reused entirely